Fix a11y-noninteractive-element-interactions

derekdang opened the pull request:
Fixes https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/no-noninteractive-element-interactions.md for #2564 

HelpModal `span` now contains a `button` for accessibility.
Removed the event.preventDefault onClick logic for `Label.svelte` and added `tabIndex` and `role` to SettingTitle.
event.preventDefault() added to openHelpModal in components with `SwitchRow` so that `SwitchRow` does not toggle on when clicking the `SettingTitle`.

I toyed around with the idea of adding a boolean `isSwitchRow` to openHelpModal if we only want to `event.preventDefault()` in ONLY that case but `event.preventDefault()` did not seem to be affecting other components upon click.

SettingTitle still has lint issues related to a11y but this PR is focused on resolving one of the four ignored warnings.

dae commented:
Thanks for working on this Derek! Your approach seems reasonable, but I've started to wonder whether we might want to explicitly ignore the warning in this case instead.

If we think about a traditional form with inputs and labels, a tabindex is not specified on the label elements. Someone using a screen reader would presumably use tab or their screen reader to jump from control to control, and their software would describe the control based on the associated label. By adding a tabindex to the labels, we've now made the common case a bit harder for everyone:

- People who are not sight impaired who want to navigate between controls now need to press tab twice to move to each control. I suspect the cases where they want to open a help link with the keyboard are likely to be rare.
- People who are sight impaired will now have the label read to them, and then have to tab to the control, where presumably the label is read again.

As the clickable labels are simply a shortcut to the relevant help section, and a separate link to the help section exists at the top of each section, perhaps it would make sense to limit the shortcut to mouse clicks, and ignore the warnings on those lines? It would mean that a11y users wouldn't be able to jump directly to a particular description, but I'm not sure that convenience justifies the inconvenience of regular navigation for them. What do you think?

derekdang commented:
> If we think about a traditional form with inputs and labels, a tabindex is not specified on the label elements. Someone using a screen reader would presumably use tab or their screen reader to jump from control to control, and their software would describe the control based on the associated label. By adding a tabindex to the labels, we've now made the common case a bit harder for everyone:
> 
> * People who are not sight impaired who want to navigate between controls now need to press tab twice to move to each control. I suspect the cases where they want to open a help link with the keyboard are likely to be rare.
> * People who are sight impaired will now have the label read to them, and then have to tab to the control, where presumably the label is read again.
> 
> As the clickable labels are simply a shortcut to the relevant help section, and a separate link to the help section exists at the top of each section, perhaps it would make sense to limit the shortcut to mouse clicks, and ignore the warnings on those lines? It would mean that a11y users wouldn't be able to jump directly to a particular description, but I'm not sure that convenience justifies the inconvenience of regular navigation for them. What do you think?

Those are good points. I do agree that the tabbing to the Labels does not mimic what is expected on a form. It is cumbersome to tab through every field twice through.

In order to better understand the approach and potential use case(s), would we want to be able to tab to the HelpModal's Badge so that it can be opened through keyboard navigation? At the moment that is not the behavior and would be addressed potentially when the a11y warnings in Badge were to be addressed and/or may require additional changes in HelpModal.

I think if users were able to tab to the Info Circle badge then that would be acceptable. I can revert the changes and ignore the warnings in Label.

With regards to the current changes in HelpModal, are we okay with the new appearance?

**Changed to Button** 
![image](https://user-images.githubusercontent.com/21960193/253377864-91c9b817-7344-40ad-9270-42d1903fb27c.png)

derekdang commented on ts/deck-options/HelpModal.svelte:
```diff
@@ -94,14 +94,17 @@ License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
-                                            <span class:active={i == activeIndex}>
-                                                {section.title}
+                                        <li>
+                                            <span>
```
Yeah, the span isn't needed. Will remove.

dae commented:
> would we want to be able to tab to the HelpModal's Badge so that it can be opened through keyboard navigation?

Yep, we'd want a11y users to be able to trigger it without a mouse, and navigate comfortably inside it too.

> are we okay with the new appearance?

I imagine with a few CSS tweaks they could be made to look more like the existing links, but that could be deferred to a later date if it's tricky, as the changed look is not particularly out of place. 

As an aside, they currently seem to change size, which causes the buttons to move from top to left or vice versa when looking at the help for the new cards section. Initially I thought this was a regression, but the old style did the same thing. Maybe setting a minimum width on them would resolve that.

adeleke5140 commented:
> Thanks for working on this Derek! Your approach seems reasonable, but I've started to wonder whether we might want to explicitly ignore the warning in this case instead.
> 
> If we think about a traditional form with inputs and labels, a tabindex is not specified on the label elements. Someone using a screen reader would presumably use tab or their screen reader to jump from control to control, and their software would describe the control based on the associated label. By adding a tabindex to the labels, we've now made the common case a bit harder for everyone:
> 
> * People who are not sight impaired who want to navigate between controls now need to press tab twice to move to each control. I suspect the cases where they want to open a help link with the keyboard are likely to be rare.
> * People who are sight impaired will now have the label read to them, and then have to tab to the control, where presumably the label is read again.
> 
> As the clickable labels are simply a shortcut to the relevant help section, and a separate link to the help section exists at the top of each section, perhaps it would make sense to limit the shortcut to mouse clicks, and ignore the warnings on those lines? It would mean that a11y users wouldn't be able to jump directly to a particular description, but I'm not sure that convenience justifies the inconvenience of regular navigation for them. What do you think?

I would like to add a comment here.
Form labels are read out to Assistive technologies in order to aid non-sighted users understand the context of the form. labels do not require the `tabIndex` attribute. The `input` elements can be tabbed to by default.

Clicking the accompanying label should bring focus to the corresponding form input.
https://github.com/ankitects/anki/assets/32024147/266a21b8-9b26-4f5d-a52d-ba56f64c34c4


Since a separate link already exists that jumps to each section, it is not necessary to have the label be used for linking.

derekdang commented:
@adeleke5140 Thanks for the insight! Yeah, that's essentially the agreement we came to.

@dae  
I made some CSS adjustments to mimic the styling prior to the changes.
Also reverted the changes and added the explicit ignore in Label.svelte.
![image](https://github.com/ankitects/anki/assets/21960193/327204cc-0a6e-410f-b4b2-7ed088c7f2d9)

dae commented:
> Since a separate link already exists that jumps to each section, it is not necessary to have the label be used for linking.

For some background on the current behaviour:

https://github.com/ankitects/anki/pull/2166

> I made some CSS adjustments to mimic the styling prior to the changes.

I've tested out the latest changes and they LGTM. Thanks Derek!

adeleke5140 commented:
Hi @dae I read through #2166 and found it to be really enlightening. Thank you for linking that.

After going through the deck options, I find the behaviour to be less than ideal.

It seems to be that the info icon should already convey that it toggles a help modal. The icon cannot be toggled with the keyboard so that should be fixed.

Modifying the form labels to also perform that action is going to be surprising for the user.

I believe you brought up those same concerns here:
https://github.com/ankitects/anki/pull/2166#issuecomment-1300004757

The implemented approach also violates Success criterion 1.1.1, 1.3.1 and possiblye 3.3.2 of WCAG 2.0 guidelines

>1.1.1 Non-text Content: All [non-text content](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#non-text-contentdef) that is presented to the user has a [text alternative](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#text-altdef) that serves the equivalent purpose, except for the situations listed below. (Level A)

Controls, Input: If non-text content is a control or accepts user input, then it has a [name](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#namedef) that describes its purpose. (Refer to [Guideline 4.1](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#ensure-compat) for additional requirements for controls and content that accepts user input.)

> 1.3.1 Info and Relationships: Information, [structure](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#structuredef), and [relationships](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#relationshipsdef) conveyed through [presentation](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#presentationdef) can be [programmatically determined](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#programmaticallydetermineddef) or are available in text. (Level A)

> 3.3.2 Labels or Instructions: [Labels](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#labeldef) or instructions are provided when content requires user input. (Level A)

What do you think?

dae commented:
I think there are valid arguments on both sides. The cursor shows a question mark when hovering over the titles, so perhaps it is not such a surprise to users that clicking on a title will show help. And the behavior only affects clicks, so theoretically should not affect users with screen readers. On the other hand, handling the click and screen reader cases separately means the code is more complicated, and it does deviate a bit from standard behavior. 

@kleinerpirat do you still feel the current behaviour is optimal, or would you be ok with dropping the clickable help titles?
